app/test/prepare_proposal_test.go

```diff
@@ -205,6 +205,61 @@ func TestPrepareProposalFiltering(t *testing.T) {
 	}
 }
 
+// TestPrepareProposalBlockDataSize verifies that prepare proposal returns a
+// response where the size of response.BlockData.Txs is less than or equal to
+// the request.BlockDataSize.
+//
+// Note: this unit test is contrived because the application's prepare proposal
+// doesn't actually enforce this constaint. Instead, celestia-core will panic if
+// the size of response.BlockData.Txs exceeds the request.BlockDataSize.
+func TestPrepareProposalBlockDataSize(t *testing.T) {
+	numBlobTxs := 1
+	numNormalTxs := 1
+	accounts := testfactory.GenerateAccounts(numBlobTxs + numNormalTxs)
+	testApp, kr := testutil.SetupTestAppWithGenesisValSet(app.DefaultConsensusParams(), accounts...)
+	encCfg := encoding.MakeConfig(app.ModuleEncodingRegisters...)
+	infos := queryAccountInfo(testApp, accounts, kr)
+
+	blobTxs := blobfactory.ManyMultiBlobTx(
+		t,
+		encCfg.TxConfig,
+		kr,
+		testutil.ChainID,
+		accounts[:numBlobTxs],
+		infos[:numBlobTxs],
+		testfactory.Repeat([]*blob.Blob{
+			blob.New(appns.RandomBlobNamespace(), []byte{1}, appconsts.DefaultShareVersion),
+		}, numBlobTxs),
+	)
+
+	normalTxs := testutil.SendTxsWithAccounts(
+		t,
+		testApp,
+		encCfg.TxConfig,
+		kr,
+		1000,
+		accounts[0],
+		accounts[numBlobTxs:],
+		testutil.ChainID,
+	)
+
+	txs := blobTxs
+	txs = append(txs, coretypes.Txs(normalTxs).ToSliceOfBytes()...)
+
+	request := abci.RequestPrepareProposal{
+		BlockData: &tmproto.Data{
+			Txs: txs,
+		},
+		BlockDataSize: getSize(txs),
+		ChainId:       testutil.ChainID,
+		Height:        testApp.LastBlockHeight() + 1,
+		Time:          time.Now(),
+	}
+
+	got := testApp.PrepareProposal(request)
+	require.LessOrEqual(t, getSize(got.BlockData.Txs), request.BlockDataSize)
```

Comment:
[not blocking]
Does it make sense to have a test where the size is bigger than the BlockDataSize? and check if prepare proposal fails?

Reply:
Unfortunately it won't fail unless we add the explicit check proposed in the PR description.